appmap/_implementation/detect_enabled.py

```diff
@@ -0,0 +1,182 @@
+"""
+Detect if AppMap is enabled.
+"""
+
+import importlib
+import logging
+import os
+from textwrap import dedent
+
+logger = logging.getLogger(__name__)
+
+
+class UnrecognizedRecordingMethodException(Exception):
+    pass
+
+
+RECORDING_METHODS = ["pytest", "unittest", "remote", "requests"]
+
+# Detects whether AppMap recording should be enabled. This test can be
+# performed generally, or for a particular recording method. Recording
+# can be enabled explicitly, for example via APPMAP=true, or it can be
+# enabled implicitly, by running in a dev or test web application
+# environment. Recording can also be disabled explicitly, using
+# environment variables.
+class DetectEnabled:
+    _instance = None
+
+    def __new__(cls):
+        if cls._instance is None:
+            logger.debug("Creating the DetectEnabled object")
+            cls._instance = super(DetectEnabled, cls).__new__(cls)
+            cls._instance._initialized = False
+        return cls._instance
+
+    def __init__(self):
+        if self._initialized:
+            return
+
+        self._initialized = True
+        self._detected_for_method = {}
+
+    @classmethod
+    def initialize(cls):
+        cls._instance = None
+        # because apparently __new__ and __init__ don't get called
```

Comment:
Right, because you're not actually asking to create an instance, i.e. you don't say `DetectEnabled()` anywhere.

Reply:
It looks to me like it gets set. Maybe I don't understand what you mean?

```
diff --git a/appmap/_implementation/detect_enabled.py b/appmap/_implementation/detect_enabled.py
index 6d03c7b..aeea80c 100644
--- a/appmap/_implementation/detect_enabled.py
+++ b/appmap/_implementation/detect_enabled.py
@@ -38,11 +38,12 @@ class DetectEnabled:
     def initialize(cls):
         cls._instance = None
         # because apparently __new__ and __init__ don't get called
-        cls._detected_for_method = {}
+        # cls._detected_for_method = {}

     @classmethod
     def clear_cache(cls):
-        cls._detected_for_method = {}
+        # cls._detected_for_method = {}
+        pass

     @classmethod
     def is_appmap_repo(cls):
sw/feat/record_by_default *$
ajp@Alans-MacBook-Pro appmap-python-2 % python
Python 3.10.7 (main, Sep 15 2022, 04:53:36) [Clang 13.1.6 (clang-1316.0.21.2.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from appmap._implementation.detect_enabled import DetectEnabled
>>> de = DetectEnabled()
>>> print(de._detected_for_method)
{}
>>>
```

Reply:
If I apply the first diff -- if I comment out `cls._detected_for_method = {}` in `initialize` -- the Flask testcases don't pass
```
poetry run pytest appmap/test/test_flask.py
```

Reply:
Well, ok, but that's not (necessarily) the same thing as "__init__ doesn't set a property of self", which is what it sounds like you meant.